src/database/transaction.rs

```diff
@@ -1,87 +1,180 @@
+    // Filter out any identities that are already in the `identities` table
+    let mut filtered_identities = vec![];
+    for identity in identities {
+        if tx
+            .get_identity_leaf_index(&identity.commitment)
+            .await?
+            .is_some()
+        {
+            tracing::warn!(?identity.commitment, "Duplicate identity");
+            tx.remove_unprocessed_identity(&identity.commitment).await?;
+        } else {
+            filtered_identities.push(identity.commitment);
+        }
+    }
+
+    let _guard = pending_insertions_mutex.lock().await;
+
+    let next_db_index = tx.get_next_leaf_index().await?;
+    let next_leaf = latest_tree.next_leaf();
+
+    assert_eq!(
+        next_leaf, next_db_index,
+        "Database and tree are out of sync. Next leaf index in tree is: {next_leaf}, in database: \
+         {next_db_index}"
+    );
+
+    let data = latest_tree.append_many(&filtered_identities);
+
+    assert_eq!(
+        data.len(),
+        filtered_identities.len(),
+        "Length mismatch when appending identities to tree"
+    );
 
-            tx.execute(update_previous_roots).await?;
+    let items = data.into_iter().zip(filtered_identities);
 
+    for ((root, _proof, leaf_index), identity) in items {
+        tx.insert_pending_identity(leaf_index, &identity, &root)
+            .await?;
+
+        tx.remove_unprocessed_identity(&identity).await?;
+    }
+
+    Ok(())
+}
+
+/// impl block for database transactions
+impl Database {
+    /// Marks the identities and roots from before a given root hash as mined
+    /// Also marks following roots as pending
+    #[instrument(skip(self), level = "debug")]
+    pub async fn mark_root_as_processed_tx(&self, root: &Hash) -> Result<(), Error> {
+        retry_tx!(self.pool, tx, mark_root_as_processed(&mut tx, root).await).await
+    }
+
+    /// Marks the identities and roots from before a given root hash as mined
+    /// Also marks following roots as pending
+    #[instrument(skip(self), level = "debug")]
+    pub async fn mark_root_as_processed_and_delete_batches_tx(
+        &self,
+        root: &Hash,
+    ) -> Result<(), Error> {
+        retry_tx!(self.pool, tx, {
+            mark_root_as_processed(&mut tx, root).await?;
+            tx.delete_batches_after_root(root).await?;
             Ok(())
         })
         .await
     }
+
+    /// Marks the identities and roots from before a given root hash as
+    /// finalized
+    #[instrument(skip(self), level = "debug")]
+    pub async fn mark_root_as_mined_tx(&self, root: &Hash) -> Result<(), Error> {
+        retry_tx!(self.pool, tx, mark_root_as_mined(&mut tx, root).await).await
+    }
+
+    #[instrument(level = "info", skip_all)]
+    pub async fn insert_identities_batch_tx(
+        &self,
+        latest_tree: &TreeVersion<Latest>,
+        identities: Vec<UnprocessedCommitment>,
+        pending_insertions_mutex: &Mutex<()>,
+    ) -> anyhow::Result<()> {
+        retry_tx!(
```

Comment:
if the underlying function fails we'll still end up in an invalid state because we'll have update the latest_tree, no? So this won't fix the next_leaf_idx desync

Reply:
Yeah you're right. Somehow I forgot about the interior mutability here. I think something like this with two transactions should be a reasonable solution:

```rust
pub async fn insert_identities_batch(
    database: &Database,
    latest_tree: &TreeVersion<Latest>,
    identities: &[UnprocessedCommitment],
    pending_insertions_mutex: &Mutex<()>,
) -> anyhow::Result<()> {
    let filtered_identities = retry_tx!(database, tx, {
        // Filter out any identities that are already in the `identities` table
        let mut filtered_identities = vec![];
        for identity in identities {
            if tx
                .get_identity_leaf_index(&identity.commitment)
                .await?
                .is_some()
            {
                tracing::warn!(?identity.commitment, "Duplicate identity");
                tx.remove_unprocessed_identity(&identity.commitment).await?;
            } else {
                filtered_identities.push(identity.commitment);
            }
        }
        Result::<_, anyhow::Error>::Ok(filtered_identities)
    })
    .await?;

    let _guard = pending_insertions_mutex.lock().await;

    let data = latest_tree.append_many(&filtered_identities);

    assert_eq!(
        data.len(),
        filtered_identities.len(),
        "Length mismatch when appending identities to tree"
    );

    let next_leaf = latest_tree.next_leaf();

    retry_tx!(database, tx, {
        let next_db_index = tx.get_next_leaf_index().await?;

        assert_eq!(
            next_leaf, next_db_index,
            "Database and tree are out of sync. Next leaf index in tree is: {next_leaf}, in \
             database: {next_db_index}"
        );

        for ((root, _proof, leaf_index), identity) in data.iter().zip(&filtered_identities) {
            tx.insert_pending_identity(*leaf_index, identity, root)
                .await?;

            tx.remove_unprocessed_identity(identity).await?;
        }

        Ok(())
    })
    .await
}
```

What do you think?